numpy/_core/src/umath/legacy_array_method.c

```diff
@@ -427,11 +420,67 @@ PyArray_NewLegacyWrappingArrayMethod(PyUFuncObject *ufunc,
     };
 
     PyBoundArrayMethodObject *bound_res = PyArrayMethod_FromSpec_int(&spec, 1);
+
     if (bound_res == NULL) {
         return NULL;
     }
     PyArrayMethodObject *res = bound_res->method;
+
+    // set cached initial value for numeric reductions to avoid creating
+    // a python int in every reduction
+    if (PyTypeNum_ISNUMBER(bound_res->dtypes[0]->type_num)) {
+        char *initial =
+                PyMem_Calloc(1, bound_res->dtypes[0]->singleton->elsize);
+
+        if (initial == NULL) {
+            PyErr_NoMemory();
+            return NULL;
+        }
+
+        PyArray_Descr **descrs = PyMem_Calloc(ufunc->nin + ufunc->nout,
+                                              sizeof(PyArray_Descr *));
+
+        if (descrs == NULL) {
+            PyErr_NoMemory();
+            PyMem_Free(initial);
+            return NULL;
+        }
+
+
+        for (int i = 0; i < ufunc->nin + ufunc->nout; i++) {
+            // only dealing with numeric legacy dtypes so this should always be
+            // valid
+            descrs[i] = bound_res->dtypes[i]->singleton;
+        }
+
+        PyArrayMethod_Context context = {
+                (PyObject *)ufunc,
+                bound_res->method,
+                descrs,
+        };
+
+        int ret = get_initial_from_ufunc(&context, 0, initial);
+
+        if (ret < 0) {
+            PyMem_Free(initial);
+            PyMem_Free(descrs);
+            return NULL;
+        }
+
+        // only use the initial value if it's valid
+        if (ret > 0) {
+            memcpy(context.method->legacy_initial, initial,
+                   context.descriptors[0]->elsize);
+            context.method->get_reduction_initial = &copy_cached_initial;
+        }
+
+        PyMem_Free(initial);
+        PyMem_Free(descrs);
+    }
+
+
     Py_INCREF(res);
     Py_DECREF(bound_res);
+
```

Comment:
`bound_res` isn't decref'ed in the errors above.  But once we remove all of the uneeded allocations, at least there is only one error path left.